src/monotonic_path_preview/mod.rs

```diff
@@ -0,0 +1,911 @@
+    /// components to escape whatever prefix it's joined to.
+    ///
+    /// [`Relative::new()`]: struct.Relative.html#method.new
+    RelativePathEscapesPrefix(path::PathBuf),
+
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        match self {
+            Error::IoError { err, at } => write!(f, "{}: {:?}", err, at),
+            Error::SymlinkLoops(p) => {
+                write!(f, "Found an infinite symlink loop: {:?}", p)
+            }
+            Error::PathIsAbsolute(p) => write!(
+                f,
+                "Tried to make a relative path from absolute path {:?}",
+                p
+            ),
+            Error::RelativePathEscapesPrefix(p) => write!(
+                f,
+                "Tried to make a relative path with leading '..': {:?}",
+                p
+            ),
+            _ => write!(f, "Unknown error"),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match self {
+            Error::IoError { err: e, at: _ } => e.description(),
+            Error::SymlinkLoops(_) => "Found an infinite symlink loop",
+            Error::PathIsAbsolute(_) => {
+                "Tried to make a relative path from an absolute path"
+            }
+            Error::RelativePathEscapesPrefix(_) => {
+                "Tried to make a relative path with leading '..'"
+            }
+            _ => "Unknown error",
+        }
+    }
+}
+
+/// Splits a path into a head and a tail.
+///
+/// The "head" consists of the Prefix and RootDir components, if any,
+/// and is fully canonicalized.
+///
+/// The "tail" consists of all the other components that follow the head.
+///
+/// # Errors
+///
+/// Returns [`Error::IoError`] if the head cannot be canonicalized.
+/// For example, if the process' current working directory has been deleted,
+/// or the given path includes a syntactically invalid prefix.
+///
+/// [`Error::IoError`]: struct.Error.html#variant.IoError
+fn split_head_and_tail<P: AsRef<path::Path>>(
+    path: P,
+) -> Result<
+    (
+        path::PathBuf,
+        collections::VecDeque<ffi::OsString>,
+    ),
+    Error,
+> {
+    let path = path.as_ref();
+
+    debug!("Splitting head and tail of {:?}", path);
+
+    // The path's head is the prefix and root components (if any).
+    fn is_head_part(c: &path::Component) -> bool {
+        match c {
+            path::Component::Prefix(_) => true,
+            path::Component::RootDir => true,
+            _ => false,
+        }
+    }
+
+    let mut head: path::PathBuf = path.components()
+        .take_while(is_head_part)
+        .collect();
+
+    // If this path is purely relative, it's relative to our current
+    // directory.
+    if head.as_os_str() == "" {
+        head.push(".");
+    }
+
+    debug!("Raw head: {:?}", head);
+
+    // The path's head can be safely converted into an absolute
```

Comment:
unfortunately this isn't true for windows. See my implementation in `arc::handle_prefix` (which is messy and poor). Windows is pure evil on this.

Here's the rust issue I opened: https://github.com/rust-lang/rust/issues/49342

Reply:
I've been using [the official Win10 test VM](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/) for VirtualBox—it's supposed to be for testing browsers, but it's a full Win10 install so you can mount your home directory and run binaries and whatever. The only catch is that it only runs for 90 days, but in VirtualBox you can take a system snapshot before the first boot and just revert back to get another 90 days without having to download a fresh image.

Can you point me to an Appveyor build that failed with this problem?